src/shims/unix/socket.rs

```diff
@@ -211,19 +280,42 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let socketpair_0 = SocketPair {
             writebuf: Rc::downgrade(&buffer1),
             readbuf: Rc::clone(&buffer2),
+            peer_fd: None,
             is_nonblock: is_sock_nonblock,
+            epoll_events: Vec::new(),
         };
-
         let socketpair_1 = SocketPair {
             writebuf: Rc::downgrade(&buffer2),
             readbuf: Rc::clone(&buffer1),
+            peer_fd: None,
             is_nonblock: is_sock_nonblock,
+            epoll_events: Vec::new(),
         };
+        let file_descriptor0 = FileDescriptor::new(socketpair_0);
+        let file_descriptor1 = FileDescriptor::new(socketpair_1);
 
         let fds = &mut this.machine.fds;
-        let sv0 = fds.insert_fd(FileDescriptor::new(socketpair_0));
+
+        let sv0 = fds.insert_fd(file_descriptor0);
+        let sv1 = fds.insert_fd(file_descriptor1);
+
+        // Expose peer file descriptor to each other.
+        if let Some(sv0_file_descriptor) = this.machine.fds.dup(sv0) {
+            let sv0_weak = sv0_file_descriptor.get_weak_file_descriptor();
+            let file_description0 = sv0_weak.get_file_description().unwrap();
+            let mut fd_borrow0 = file_description0.borrow_mut();
+            let sv0_file_description = fd_borrow0.downcast_mut::<SocketPair>().unwrap();
+            if let Some(sv1_file_descriptor) = this.machine.fds.dup(sv1) {
+                let sv1_weak = sv1_file_descriptor.get_weak_file_descriptor();
+                let file_description1 = sv1_weak.get_file_description().unwrap();
+                let mut fd_borrow1 = file_description1.borrow_mut();
+                let sv1_file_description = fd_borrow1.downcast_mut::<SocketPair>().unwrap();
+                sv0_file_description.add_peer_fd(sv1_weak);
+                sv1_file_description.add_peer_fd(sv0_weak);
+            };
+        };
```

Comment:
Hmm this is the only way that I could think of to expose the file descriptor. I still couldn't find a way to construct a ``Weak`` before inserting the fd to fd table without the compiler yelling at me... But this attempt looks too verbose.